kore/src/Kore/JsonRpc.hs

```diff
@@ -527,30 +527,27 @@ respond serverState moduleName runSMT =
                         parseKoreModule "<add-module>" _module
             st@ServerState
                 { serializedModules
+                , receivedModules
                 , loadedDefinition = LoadedDefinition{indexedModules, definedNames, kFileLocations}
                 } <-
                 liftIO $ MVar.takeMVar serverState
             let moduleHash = ModuleName . fromString . ('m' :) . show . hashWith SHA256 $ encodeUtf8 _module
 
             -- put the original state back if we fail at any point
             flip catchE (\e -> liftIO (MVar.putMVar serverState st) >> throwError e) $ do
+                -- check if we already received a module with this name
                 when nameAsId $
-                    case (Map.lookup (coerce name) indexedModules, Map.lookup (coerce moduleHash) indexedModules) of
-                        (Just{}, Nothing) ->
-                            -- another module with the same name already exists
-                            throwError $ backendError DuplicateModuleName name
-                        (Just nmMod, Just idMod)
-                            | nmMod /= idMod ->
-                                -- this module has previously been added and different
-                                -- module with the same name also already exists
-                                throwError $ backendError DuplicateModuleName name
-                            | otherwise ->
-                                -- this module has previously been added with name-as-id: true
-                                -- we can allow this, since the contents of the named module
-                                -- are the same
-                                pure ()
+                    case Map.lookup (coerce name) receivedModules of
+                        -- if a different module was already added, throw error
+                        Just m | _module /= m -> throwError $ backendError DuplicateModuleName name
                         _ -> pure ()
 
+                -- Check for a corner case when we send module M1 with the name "m<hash of M2>"" and name-as-id: true
+                -- followed by adding M2. Should not happen in practice...
+                case Map.lookup (coerce moduleHash) receivedModules of
+                    Just m | _module /= m -> throwError $ backendError DuplicateModuleName name
```

Comment:
Of course we could return `coerce moduleHash` in the error here...

Reply:
Yeah, I dunno. There doesn't seem to be a good answer there... probably better to return the module hash indeed, tho this error would be very confusing if it ever happened...again these are rather philosophical points...